### ISSUE TITLE
WSLg doesn't pick up the actual icons of Linux desktop apps

### ISSUE THREAD

**Issue by bbatsov** [microsoft/wslg#614]
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.22000.348]
Your Distribution version:  Ubuntu 20.04
Your WSLg version: 1.0.26
```

# Steps to reproduce

Just run any Linux GUI from the command line or the Windows start menu. All apps have the generic Linux logo icon, which makes them hard to differentiate.

I've installed all apps with `apt install`. 

#  Expected behavior

I expect each app to display it's all icon (as defined by `appname.desktop`). 

# Actual behavior

![image](https://user-images.githubusercontent.com/103882/146922123-a354c304-fb21-4d91-b4ad-c872e221c06e.png)

The same generic icon is displayed in the taskbar as well. Interestingly, when I run Emacs as `root` it displays the right icon. 

![image](https://user-images.githubusercontent.com/103882/146922334-238121c9-2845-4224-bc2c-275dd4abbb52.png)

**Comment by tcfurrer** [comment]
Exact same issue here.
I would like to add that the correct icon shows up separately in the taskbar when the app is running, but the pinned icon and start menu icon is incorrect as already described here.
![image](https://user-images.githubusercontent.com/39924916/147371010-94be2d94-9f99-4d5f-bb83-9a23e082b31f.png)

**Comment by seven-mile** [comment]
I've been suffering from this issue for months. I guess it relates to the Ubuntu version (mine is also 20.04).

**Comment by bbatsov** [comment]
>  I would like to add that the correct icon shows up separately in the taskbar when the app is running, but the pinned icon and start menu icon is incorrect as already described here.

For me the icon was the correct one initially, but after a while all icons from Ubuntu got replaced with the generic one.

**Comment by hideyukn88** [comment]
@bbatsov, @tcfurrer, @seven-mile, would you please share /mnt/wslg/weston.log with us? 

Basically there is 2 type of icons, a) icons used in Start Menu (or pinned at taskbar), and b) icons at taskbar when applications are running. 

Icon a) is specified at .desktop file for that application, for emacs, typically /usr/share/applications/emacs.desktop, which specify as "Icon=emacs", then Icon file will be loaded from /usr/share/icons. And if no icon is specified at .desktop file, then generic Icon will be used.

![image](https://user-images.githubusercontent.com/13857962/148142229-1f34367b-eb1b-49ea-bb2b-1f9d721dd04c.png)

Icon b) is from "_NET_WM_ICON" for X11 application. So if you are not seeing right Icon here, would you check by running xprop ? Please refer below screen capture. And in case application is Wayland native application, same Icon as a) will be used.

![image](https://user-images.githubusercontent.com/13857962/148141991-0f70b4a1-5e89-4722-8cb8-35ff18f6dbe9.png)

Thanks!

**Comment by seven-mile** [comment]
@hideyukn88 Of course. For (a) icons, on my machine, only gThumb is normal but others are generic icon. Indeed the .desktop configurations correspond with this fact.
![nautilus's .desktop file](https://cdn.luogu.com.cn/upload/image_hosting/nwb9m8m1.png)
![gThumb's .desktop file](https://cdn.luogu.com.cn/upload/image_hosting/mu8sjedl.png)

But the (b) icons of all applications including nautilus are not working in the taskbar. I tried `xprop` `wmctrl` and `xwininfo`. But it seemed none of them worked. You may wait for feedback from others.

Attachment:
[weston.log.zip](https://github.com/microsoft/wslg/files/7812176/weston.log.zip)

**Comment by hideyukn88** [comment]
@seven-mile, thanks for info.

First, for icon a), from your weston.log, the most of your application is using SVG format icon, which is typical in many gnome application, so this is not yet supported. Please see https://github.com/microsoft/wslg/issues/70 and we are working on it (https://github.com/microsoft/CBL-MarinerCoreUI/pull/95)

```
[13:13:52.657]     Icon file:/usr/share/icons/hicolor/scalable/apps/org.gnome.Nautilus.svg
[13:13:52.658]     Icon file:/usr/share/icons/hicolor/scalable/apps/org.gnome.gedit.svg
[13:13:52.659]     Icon file:/usr/share/icons/hicolor/scalable/apps/org.gnome.Evince.svg
```

For icon b), if those gnome application is running as Wayland native mode, then without SVG support, it won't be able to obtain Icon for taskbar. You can try running with in X11 mode by "export GDK_BACKEND=x11", then launch your application, please let us know if this helps. Thanks!

**Comment by seven-mile** [comment]
@hideyukn88 
Yes, "export GDK_BACKEND=x11" works well. Looking forward to SVG support in this year!

**Comment by hideyukn88** [comment]
@bbatsov, @tcfurrer, I would be interested in the missing icons for Emacs case, if you are still experiencing this issue, please let me know. Thanks!

**Comment by bbatsov** [comment]
Sure, I've attached my `weston.log`. 

[weston.log](https://github.com/microsoft/wslg/files/7827178/weston.log)

I notice in it things like:

![image](https://user-images.githubusercontent.com/103882/148507548-5bf38841-9bf5-46c8-ab93-350a19ef4f2d.png)

And I had to do a screenshot as copying from WSLg doesn't work for me either. :-)

The Emacs desktop file is basically the default one and the icon is where it should be. I'm not sure what I'm supposed to do with `xprop`, so I'll take at look at this later.

**Comment by kharus** [comment]
It looks like weston doesn't look for icons in `/usr/local/share/icons/`
As a workaround for emacs I edited file
`/usr/local/share/applications/emacs.desktop`
and changed Icon to `emacs29` and copied icon files from emacs checkout to `/usr/share/icons`
```
cd $EMACS_CODE/etc/images/icons/hicolor/
for d in *                                                                                                                                                                                                                            
do
sudo cp $d/apps/emacs.png /usr/share/icons/hicolor/$d/apps/emacs29.png
done
```
Restarted wslg and emacs now has emacs icon in windows.

**Comment by seflerZ** [comment]
This is because current WSLg doesn't support SVG icon. Though this function code has been added but the corresponding SVG lib can't be found in Mariner (A Microsfoft linux distro) repository. Refer to [this ](https://github.com/microsoft/CBL-Mariner/issues/1599#issuecomment-958290924) to has more details.

**Comment by YAmikep** [comment]
I am having the same issue and could not resolve it reading the above.
For example, I installed Gitkraken:
```
wget https://release.gitkraken.com/linux/gitkraken-amd64.deb
sudo apt install ./gitkraken-amd64.deb
```

Then I launch it:
```
gitkraken
```

I see the generic icon.

Here are the logs:
```
❯ cat /mnt/wslg/weston.log

[13:44:37.885] set_position_from_xwayland: XWM 860, 302; geometry 32, 32; view 828, 270
[13:44:37.901] app_list_monitor_thread: loadIconEvent is signalled. GitKraken
[13:44:37.901] app_list_monitor_thread: entry (nil), image (nil)
[13:44:37.937] Client: ClientGetAppidReq: pid:9707 appId:GitKraken
[13:44:40.489] set_position_from_xwayland: XWM 334, 140; geometry 32, 32; view 302, 108
[13:44:40.548] app_list_monitor_thread: loadIconEvent is signalled. GitKraken
[13:44:40.548] app_list_monitor_thread: entry (nil), image (nil)
[13:44:40.649] Client: ClientGetAppidReq: pid:9707 appId:GitKraken
``

Any idea how to fix this?

**Comment by YAmikep** [comment]
Another use case is when using an AppImage.
In that case, there is no entry in `/usr/share/applications/` so it's not a mismatch of AppId.
Any workaround to this?

Thanks

**Comment by TheAlienDrew** [comment]
Hey guys, at least for me using Ubuntu, running `sudo update-icon-caches /usr/share/icons/*` seems to fix the icons for me.

**Comment by andnig** [comment]
I have a similar issue, but none of the workarounds helped so far. 
All the icons are correctly set in the start menu, however not in the task bar.
![image](https://github.com/microsoft/wslg/assets/30227237/29e34afa-1fef-44de-af35-10915f2791b5)

Task bar:
![image](https://github.com/microsoft/wslg/assets/30227237/23217bcd-6b72-4547-bcac-1782a8571573)

WSL version: 2.0.14.0
WSLg version: 1.0.59

For eg. for kitty, my desktop file is at `/usr/share/applications/kitty.desktop`, Icon has an absolute path to a .png file. (I guess that's why it's correctly showing in the start menu?)

Is there a chance to get the icons also working on the task bar? Where do we need to put the icons so that they are used in the taskbar?

**Comment by lxg208** [comment]
This problem has puzzled me for a while and I think I found a solution. The reason is missing .desktop file. On a new machine I installed Ubuntu 23.10. I compiled emacs from the source and it suffers the same issue. Then I created the missing .desktop file and I have the emacs icon. Back to wsl, I created the same desktop file and it worked as well. Following is my /usr/share/applications/emacs.desktop file, notice the location of emacs and icon file:

```
[Desktop Entry]
Version=1.0
Name=Emacs (GUI)
GenericName=Text Editor
Comment=GNU Emacs is an extensible, customizable text editor - and more
MimeType=text/english;text/plain;text/x-makefile;text/x-c++hdr;text/x-c++src;text/x-chdr;text/x-csrc;text/x-java;text/x-moc;text/x-pascal;text/x-tcl;text/x-tex;application/x-shellscript;text/x-c;text/x-c++;
TryExec=/home/myuser/local/emacs/bin/emacs
Exec=/home/myuser/local/emacs/bin/emacs %F
Icon=/home/myuser/local/emacs/share/icons/hicolor/scalable/apps/emacs.svg
Type=Application
Terminal=false
Categories=Utility;Development;TextEditor;
StartupWMClass=Emacs
Keywords=Text;Editor;

```

**Comment by terlar** [comment]
Thank you very much @andnig after, following the cache refresh tip from that MS answer I finally got the icons works for the task bar as well.

**Comment by JotaRata** [comment]
I have the same issue and so far nothing has worked for me but I noticed that the correct icons are displayed in the taskbar when running the app with sudo.

![imagen](https://github.com/user-attachments/assets/818eff34-cbb0-42a4-8289-d3d438ddbd0e)

Running the same app without using sudo makes the taskbar to show the default icon

**Comment by noepicard** [comment]
Also make sure that the .desktop file is named the same as the `StartupWMClass`.

For example, `~/.local/share/applications/jetbrains-idea.desktop`:
```
[Desktop Entry]
...
Icon=/home/foo/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate/bin/idea.svg
StartupWMClass=jetbrains-idea
```

**Comment by Tzrlk** [comment]
In case it helps anyone, I made an overly-complicated script to automatically convert all your desktop files to use .ico instead of .svg icons:
https://gist.github.com/Tzrlk/aaee860b0602d2592bfd4e171b5ffb28

No support for /usr/share/icons yet, but I'll put that together if there's demand.
